[qos] WRRtest: pace egress with a shaper instead of scale_factor tolerance

pinky-nexthop opened the pull request:
### Description of PR

Make `WRRtest` robust to ptf-side packet loss on high-speed ports.

Original approach in this PR tolerated the loss with a `scale_factor`.
Per review feedback, this has been reworked to **pace the dst port with a
reusable egress shaper** instead, so the released burst does not overrun the
fanout→PTF path and the strict per-queue checks stay strict.

- `apply_egress_shaper()` is armed on the dst port while it is still
  tx-enabled (before `tx_disable`) and cleared in `finally`. The helper is
  Broadcom-only (no-op otherwise) and sets only the **port-level** scheduler,
  leaving the per-queue DWRR schedulers — the weight arbitration under test —
  untouched.
- Rate defaults to 1 Gbps (validated for zero ptf-side loss on Broadcom HW),
  overridable per-yaml via `egress_shaper_rate` in the `wrr`/`wrr_chg` block.
- Per-dscp bound relaxed to `diff <= limit`.

**Depends on #25453**, which introduces the reusable
`apply_egress_shaper()` / `clear_egress_shaper()` base-class helpers in
`sai_base_test.py`. This PR should be rebased once #25453 merges — the helpers
then come from master and no code is duplicated here. Kept as **draft** until
then.

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [x] Test case improvement

### Approach
Reuse the egress-shaper framework (from #25453) for the DWRR testlet in place
of the `scale_factor` loss tolerance.

mssonicbld commented:
/azp run

pinky-nexthop commented:
@vmittal-msft 
Hey Vineet, could you please have a look and merge the above code. Any suggestions are welcome. Thanks.

vmittal-msft reviewed (changes requested):
@pinky-nexthop please share possible scenarios causing this behavior. Usually it happens when fanout is not able to take burst at 400g and drops packets on way to PTF. We should first try to optimize fanout path to see if we can get 100%. Also, the change shouldn't be for all platforms but only UT2.

pinky-nexthop commented:
> @pinky-nexthop please share possible scenarios causing this behavior. Usually it happens when fanout is not able to take burst at 400g and drops packets on way to PTF. We should first try to optimize fanout path to see if we can get 100%. Also, the change shouldn't be for all platforms but only UT2.

Thanks for taking a look
Saw an issue with higher port speeds >= 400G, where few packets are lost and the test fails. 
You are right, it happens when fanout is unable to take the burst at higher speeds. This change is backward compatible in the sense that if 100% of packets are received, verification is done with 1.0 scale factor but if couple of packets are lost (defined by scale factor in the yaml file. if set to 0.95, then test can lose 0.05 packets), and still do the verification instead of failing the test.  
I tried to come up with a better solution that proposed in the below open PR:
https://github.com/sonic-net/sonic-mgmt/pull/15353/files

Also, why this should be applicable to UT2 only?

pinky-nexthop commented:
> @pinky-nexthop please list action items from the meeting here so we can track the work,

We had an offline discussion and Vineet suggested to use a shaper to avoid high speed burst at the fanout from the dut. This will make sure that all the packets are received at the fanout and hence the ptf. 
I need to try these changes and update the PR

rlhui commented:
> > @pinky-nexthop please list action items from the meeting here so we can track the work,
> 
> We had an offline discussion and Vineet suggested to use a shaper to avoid high speed burst at the fanout from the dut. This will make sure that all the packets are received at the fanout and hence the ptf. I need to try these changes and update the PR

@pinky-nexthop , wondering if this is done? thanks

pinky-nexthop commented:
> > > @pinky-nexthop please list action items from the meeting here so we can track the work,
> > 
> > 
> > We had an offline discussion and Vineet suggested to use a shaper to avoid high speed burst at the fanout from the dut. This will make sure that all the packets are received at the fanout and hence the ptf. I need to try these changes and update the PR
> 
> @pinky-nexthop , wondering if this is done? thanks

Not yet. I will move this to draft until done. thanks.

pinky-nexthop committed:
[qos] WRRtest: pace egress with a shaper instead of scale_factor tolerance

Address review feedback on this PR: instead of tolerating ptf-side packet
loss with a scale_factor, pace the dst port with the reusable egress shaper
so the released burst does not overrun the fanout->PTF path and the strict
per-queue checks can stay strict.

Apply apply_egress_shaper() on the dst port while it is still tx-enabled
(armed before the burst is released) and clear it in the finally. The helper
is broadcom-only (no-op otherwise) and only sets the port-level scheduler, so
the per-queue DWRR schedulers -- the weight arbitration under test -- are
untouched. Rate is overridable per-yaml via egress_shaper_rate in the
wrr/wrr_chg block (default 1 Gbps, validated for zero ptf-side loss on
Broadcom hardware). The per-dscp bound is relaxed to diff <= limit.

The reusable apply_egress_shaper()/clear_egress_shaper() base-class helpers
are introduced by #25453; this change depends on that PR and should be
rebased once it merges.

Signed-off-by: Pinky Agrawal <pinky@nexthop.ai>

mssonicbld commented:
/azp run

mssonicbld commented:
This PR has backport request label(s) for branch(es): msft-202503, but is missing required test information. Please make sure you tick the tested branch(es) in the **Tested branch** section and provide test evidence (e.g., 202503: &lt;test result&gt;) in the **Test result** section as well in your PR description.

<div align="right"><sub>---Powered by SONiC BuildBot</sub></div>
